File: src/main/java/pro/tuscan/domain/player/PlayerPerformanceMapper.java
package pro.tuscan.domain.player;

import pro.tuscan.adapter.api.response.dto.MapStats;
import pro.tuscan.adapter.api.response.dto.Performance;
import pro.tuscan.adapter.api.response.dto.SoloPerformance;
import pro.tuscan.adapter.api.response.dto.TeamPerformance;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class PlayerPerformanceMapper {

    private PlayerPerformanceMapper() {
    }

    public static Performance map(List<MapStats> mapStats) {
        return new Performance(
                new SoloPerformance(orderFromHighestKdRatio(mapStats).get(0).getCsgoMap(),
                                    orderFromHighestKdRatio(mapStats).get(0).getKdRatio()),
                new TeamPerformance(
                        orderFromHighestWinRate(mapStats).get(0).getCsgoMap(),
                        orderFromHighestWinRate(mapStats).get(0).getWinPercentage()
                ),
                new SoloPerformance(orderFromLowestKdRatio(mapStats).get(0).getCsgoMap(),
                        orderFromLowestKdRatio(mapStats).get(0).getKdRatio()),
                new TeamPerformance(
                        orderFromLowestWinRate(mapStats).get(0).getCsgoMap(),
                        orderFromLowestWinRate(mapStats).get(0).getWinPercentage()
                ));
    }

    static List<MapStats> orderFromHighestKdRatio(List<MapStats> mapStats) {
        return mapStats.stream()
                .sorted(Comparator.comparing(MapStats::getKdRatio).reversed())
                .collect(Collectors.toList());
    }

    private static List<MapStats> orderFromHighestWinRate(List<MapStats> mapStats) {
        return mapStats.stream()
                .sorted(Comparator.comparing(MapStats::getWinPercentage).reversed())
                .collect(Collectors.toList());
    }

    private static List<MapStats> orderFromLowestKdRatio(List<MapStats> mapStats) {
        return mapStats.stream()
                .sorted(Comparator.comparing(MapStats::getKdRatio))
                .collect(Collectors.toList());
    }

    private static List<MapStats> orderFromLowestWinRate(List<MapStats> mapStats) {
        return mapStats.stream()
                .sorted(Comparator.comparing(MapStats::getWinPercentage))
                .collect(Collectors.toList());
    }
}

File: src/main/java/pro/tuscan/domain/player/PlayerPerformanceMapper.kt
package pro.tuscan.domain.player

import pro.tuscan.adapter.api.response.dto.MapStats
import pro.tuscan.adapter.api.response.dto.Performance
import pro.tuscan.adapter.api.response.dto.SoloPerformance
import pro.tuscan.adapter.api.response.dto.TeamPerformance

object PlayerPerformanceMapper {

    @JvmStatic
    fun map(mapStats: List<MapStats>): Performance {
        return Performance(
                bestSoloPerformance = getExtremeByKdRatio(mapStats, true)
                        .let { SoloPerformance(it.csgoMap, it.kdRatio) },
                bestTeamPerformance = getExtremeByWinRate(mapStats, true)
                        .let { TeamPerformance(it.csgoMap, it.winPercentage) },
                worstSoloPerformance = getExtremeByKdRatio(mapStats, false)
                        .let { SoloPerformance(it.csgoMap, it.kdRatio) },
                worstTeamPerformance = getExtremeByWinRate(mapStats, false)
                        .let { TeamPerformance(it.csgoMap, it.winPercentage) })
    }

    private fun getExtremeByKdRatio(mapStats: List<MapStats>, reversed: Boolean) =
            mapStats.sortedBy { it.kdRatio }
                    .let { if (reversed) it.reversed() else it }
                    .first()

    private fun getExtremeByWinRate(mapStats: List<MapStats>, reversed: Boolean) =
            mapStats.sortedBy { it.winPercentage }
                    .let { if (reversed) it.reversed() else it }
                    .first()
}
